stumpy/core.py

```diff
@@ -2559,3 +2559,95 @@ def _select_P_ABBA_value(P_ABBA, k, custom_func=None):
             MPdist = partition[k]
 
     return MPdist
+
+
+@njit(parallel=True)
+def _merge_topk_PI(PA, PB, IA, IB):
+    """
+    Merge two top-k matrix profiles PA and PB, and update PA (in place) while
+    always choosing values of PA over values of PB in case of ties. Also, update
+    IA accordingly.
+
+    Parameters
+    ----------
+    PA : numpy.ndarray
+        A (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order. Also, it needs to be the same shape as PB.
+
+    PB : numpy.ndarray
+        A (top-k) matrix profile, with ndim of 2, where values in each row are
+        sorted in ascending order. Also, it needs to be the same shape as PA.
+
+    IA : numpy.ndarray
+        A (top-k) matrix profile indices, corresponding to PA
+
+    IB : numpy.ndarray
+        A (top-k) matrix profile indices, corresponding to PB
+
+    Returns
+    -------
+    None
+    """
+    for i in prange(PB.shape[0]):
+        start = 0
+        stop = np.searchsorted(PA[i], PB[i, -1], side="right")
+
+        for j in range(PB.shape[1]):
+            if PB[i, j] < PA[i, -1]:
+                idx = np.searchsorted(PA[i, start:stop], PB[i, j], side="right") + start
+
+                _shift_insert_at_index(PA[i], idx, PB[i, j])
+                _shift_insert_at_index(IA[i], idx, IB[i, j])
+
+                start = idx
+                stop += 1  # because of shifting elements to the right by one
```

Comment:
Sure. How do you feel about this?
```
for j in range(PB.shape[1]):
    if PB[i, j] >= PA[i, -1]: # because obviously, in next iteration, we have PB[i, j+1] >= PB[i, j] 
        break
    idx = np.searchsorted(PA[i, start:stop], PB[i, j], side="right") + start
    # ##
        
```

(I had this in mind, but then I thought the if-check should be cheap. I should have tested in for 100000 iterations to get my answer!)

Reply:
> I think that we should be able to `break` out of the inner `for` loop

I think when the condition ` PB[i, j] < PA[i, -1]` is satisfied, and we go inside the if-block... we cannot do break. I mean it  will not be executed. Because we MUST update `PA[i]`. The reason is that `PB[i, j]` is less than `PA[i, -1]` since we are INSIDE the if-block already. So, we had to get rid of `PA[i, -1]` and insert  `PB[i, j]` somewhere in the array `PA[i]`.

Reply:
@seanlaw 
Also, we may further optimize as follow:
```
for i in prange(PB.shape[0]):
        start = 0
        stop = np.searchsorted(PA[i], PB[i, -1], side="right")
        
        # NEW Optimization 
        if stop == 0: # means maximum value in PB[i] is smaller than the smallest value in PA[i].
            PA[i] = PB[i]
            IA[i]  = IB[i]
            continue
        
        for j in range(PB.shape[1]):
            if PB[i, j] >= PA[i, -1]:
                break # NEW Optimization  (your suggestion) 
            idx = np.searchsorted(PA[i, start:stop], PB[i, j], side="right") + start

            _shift_insert_at_index(PA[i], idx, PB[i, j])
            _shift_insert_at_index(IA[i], idx, IB[i, j])

            start = idx
            stop += 1  # because of shifting elements to the right by one
```

Reply:
I am okay with this. As much as I dislike `continue` and `break`, it is acceptable given that the function is so simple. It would be good to keep a comment for both (new) optimizations

Reply:
> As much as I dislike `continue` and `break`

I know :) 

> it is acceptable given that the function is so simple. It would be good to keep a comment for both (new) optimizations

Cool. I will add them.